test/installers/service_installer.go

```diff
@@ -0,0 +1,151 @@
+}
+
+// UninstallServicesFromNamespace uninstalls all services from specific namespace. Returns false in case of any error while uninstalling.
+func UninstallServicesFromNamespace(namespace string) (success bool) {
+	success = true
+	if sis, ok := installedServices.Load(namespace); ok {
+		for _, si := range sis.([]*ServiceInstaller) {
+			// Remove all the CRs created in the namespace, triggered for all installed services in advance as some CRs may have other CRs as owners
+			if err := si.deleteCrsWithoutOwner(namespace); err != nil {
+				framework.GetLogger(namespace).Error(err, "Error removing CRs from namespace")
+				success = false
+			}
+		}
+		for _, si := range sis.([]*ServiceInstaller) {
+			// Wait until CRs are removed
+			if err := si.waitForAllCrsRemoval(namespace, 1*time.Minute); err != nil {
+				framework.GetLogger(namespace).Error(err, "Error waiting for removal of CRs from namespace")
+			}
+			// Delete the service itself
+			if si.uninstallFromNamespace != nil {
+				if err := si.uninstallFromNamespace(namespace); err != nil {
+					framework.GetLogger(namespace).Error(err, "Error uninstalling service from namespace")
+					success = false
+				}
+			}
+		}
+	}
+	return
+}
+
+// UninstallFromCluster uninstalls the cluster wide service from whole cluster. Returns false in case of any error while uninstalling.
+func UninstallFromCluster() (success bool) {
+	success = true
+	toUninstall := map[*ServiceInstaller]bool{}
+
+	// Get unique service installers
+	installedServices.Range(func(_, sis interface{}) bool {
+		for _, si := range sis.([]*ServiceInstaller) {
+			toUninstall[si] = true
+		}
+		return true
+	})
+
+	for si := range toUninstall {
+		if si.uninstallFromCluster != nil {
+			if err := si.uninstallFromCluster(); err != nil {
+				framework.GetMainLogger().Error(err, "Error uninstalling service from cluster.")
+				success = false
+			}
+		}
+	}
+
+	return
+}
+
+// Helper methods
+
+func (installer *ServiceInstaller) deleteCrsWithoutOwner(namespace string) error {
+	crs, err := installer.getCrsWithoutOwner(namespace)
+	if err != nil {
+		return err
+	}
+
+	for _, cr := range crs {
+		if err := framework.DeleteObject(cr); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (installer *ServiceInstaller) getCrsWithoutOwner(namespace string) (crsWithoutOwner []meta.ResourceObject, err error) {
+	crs, err := installer.getAllCrsInNamespace(namespace)
+	if err != nil {
+		return nil, err
+	}
+
+	crsWithoutOwner = []meta.ResourceObject{}
+	for _, cr := range crs {
+		if len(cr.GetOwnerReferences()) == 0 {
+			crsWithoutOwner = append(crsWithoutOwner, cr)
+		}
+	}
+	return crsWithoutOwner, nil
+}
+
+func (installer *ServiceInstaller) waitForAllCrsRemoval(namespace string, timeout time.Duration) (err error) {
+	return framework.WaitFor(namespace, "Removal of all related CRs without owner", timeout, func() (bool, error) {
+		crs, err := installer.getAllCrsInNamespace(namespace)
```

Comment:
shouldn't it be `getCrsWithoutOwner` ?

Reply:
Not necessary, this function waits until all the CRs are removed, to make sure that service can be removed without leaving there any hanging CR. Regardless of whether it is with or without owner.
Some CRs have finalizer and need to be processed by service even when there is owner defined.

Reply:
Then I guess, the `WaitFor` text just before should be change or ?

Reply:
Right, copy paste error :)